src/coreclr/jit/gentree.cpp

```diff
@@ -18720,6 +18720,41 @@ bool GenTree::IsArrayAddr(GenTreeArrAddr** pArrAddr)
     return false;
 }
 
+//------------------------------------------------------------------------
+// CanSetZeroFlag: Returns true if this is an arithmetic operation that can
+// set the "zero flag" as part of its operation.
+//
+// Return Value:
+//    True if so.
+//
+// Remarks:
+//    For example, for EQ (AND x y) 0, both xarch and arm64 can emit instructions
+//    that directly set the flags after the 'AND' and thus no comparison is needed.
+//
+bool GenTree::CanSetZeroFlag()
+{
+#if defined(TARGET_XARCH)
+    if (OperIs(GT_AND, GT_OR, GT_XOR, GT_ADD, GT_SUB, GT_NEG))
+    {
+        return true;
+    }
+
+#ifdef FEATURE_HW_INTRINSICS
+    if (OperIs(GT_HWINTRINSIC) && emitter::DoesWriteZeroFlag(HWIntrinsicInfo::lookupIns(AsHWIntrinsic())))
```

Comment:
Worth noting that this isn't necessarily "correct".

`lookupIns` catches the most common cases, but certainly not all cases particularly when various intrinsics represent complex helpers that are lowered or when the table tracked instruction is `INS_invalid` due to it requiring additional lookups/handling.

Reply:
Makes sense. I added another check to the heuristic in if-conversion for the rest of the hw intrinsic cases that this PR doesn't handle yet, so that we don't regress those.